web_api/core/views.py

```diff
@@ -133,6 +168,279 @@ def current_account(request: HttpRequest, team_id: str) -> HttpResponse:
+
+    # https://stripe.com/docs/api/checkout/sessions/create
+    session = stripe.checkout.Session.create(
+        client_reference_id=account.id,
+        customer=customer_id,
+        # cards only work with subscriptions and StripeCustomerInformation
+        # depends on credit cards
+        # (payment_method_card_{brand,exp_month,exp_year,last4}).
+        payment_method_types=["card"],
+        subscription_data={
+            "items": [{"plan": settings.STRIPE_PLAN_ID, "quantity": seat_count}],
+        },
+        success_url=f"{settings.KODIAK_WEB_APP_URL}/t/{account.id}/usage?install_complete=1",
+        cancel_url=f"{settings.KODIAK_WEB_APP_URL}/t/{account.id}/usage?start_subscription=1",
+    )
+    return JsonResponse(
+        dict(
+            stripeCheckoutSessionId=session.id,
+            stripePublishableApiKey=settings.STRIPE_PUBLISHABLE_API_KEY,
+        )
+    )
+
+
+@csrf_exempt
+@auth.login_required
+def modify_payment_details(request: HttpRequest, team_id: str) -> HttpResponse:
+    account = get_object_or_404(
+        Account.objects.filter(memberships__user=request.user), id=team_id
+    )
+    session = stripe.checkout.Session.create(
+        client_reference_id=account.id,
+        customer=account.stripe_customer_id or None,
+        mode="setup",
+        payment_method_types=["card"],
+        success_url=f"{settings.KODIAK_WEB_APP_URL}/t/{account.id}/usage?install_complete=1",
+        cancel_url=f"{settings.KODIAK_WEB_APP_URL}/t/{account.id}/usage?modify_subscription=1",
+    )
+    return JsonResponse(
+        dict(
+            stripeCheckoutSessionId=session.id,
+            stripePublishableApiKey=settings.STRIPE_PUBLISHABLE_API_KEY,
+        )
+    )
+
+
+@csrf_exempt
+@auth.login_required
+def cancel_subscription(request: HttpRequest, team_id: str) -> HttpResponse:
+    account = get_object_or_404(
+        Account.objects.filter(memberships__user=request.user), id=team_id
+    )
+    if not request.user.can_edit(account):
+        raise PermissionDenied
+
+    customer_info = account.stripe_customer_info()
+    if customer_info is not None:
+        customer_info.cancel_subscription()
+    return HttpResponse(status=204)
+
+
+@csrf_exempt
+@auth.login_required
+def fetch_proration(request: HttpRequest, team_id: str) -> HttpResponse:
+    account = get_object_or_404(
+        Account.objects.filter(memberships__user=request.user), id=team_id
+    )
+    subscription_quantity = int(request.POST["subscriptionQuantity"])
+
+    customer_info = account.stripe_customer_info()
+    if customer_info is not None:
+        proration_date = int(time.time())
+        return JsonResponse(
+            dict(
+                proratedCost=customer_info.preview_proration(
+                    timestamp=proration_date,
+                    subscription_quantity=subscription_quantity,
+                ),
+                prorationTime=proration_date,
+            )
+        )
+    return HttpResponse(status=500)
```

Comment:
is this something we expect to happen often, if so we might want to return an more structured error that we can surface in the UI

Reply:
no, this shouldn't happen normally. It doesn't make sense to modify your subscription if you don't have one.